master/pkg/searcher/asha_stopping.go

```diff
@@ -0,0 +1,259 @@
+
+	var ops []Operation
+	// If the trial has completed the top rung's validation, close the trial.
+	if rungIndex == s.NumRungs-1 {
+		rung.Metrics = append(rung.Metrics,
+			trialMetric{
+				RequestID: requestID,
+				Metric:    metric,
+			},
+		)
+
+		if !s.EarlyExitTrials[requestID] {
+			ops = append(ops, NewClose(requestID))
+			s.ClosedTrials[requestID] = true
+		}
+	} else {
+		// This is not the top rung, so do promotions to the next rung.
+		nextRung := s.Rungs[rungIndex+1]
+		promoteTrial := rung.continueTraining(
+			requestID,
+			metric,
+			s.Divisor,
+		)
+		if promoteTrial {
+			s.TrialRungs[requestID] = rungIndex + 1
+			nextRung.OutstandingTrials++
+			if !s.EarlyExitTrials[requestID] {
+				unitsNeeded := max(nextRung.UnitsNeeded.Units-rung.UnitsNeeded.Units, 1)
+				ops = append(ops, NewTrain(requestID, model.NewLength(s.Unit(), unitsNeeded)))
+				ops = append(ops, NewValidate(requestID))
+				addedTrainWorkload = true
+			} else {
+				// We make a recursive call that will behave the same
+				// as if we'd actually run the promoted job and received
+				// the worse possible result in return.
+				return s.promoteAsync(ctx, requestID, ashaExitedMetricValue)
+			}
+		} else {
+			ops = append(ops, NewClose(requestID))
+			s.ClosedTrials[requestID] = true
+		}
+	}
+
+	allTrials := len(s.TrialRungs) - s.InvalidTrials
+	if !addedTrainWorkload && allTrials < s.MaxTrials {
+		create := NewCreate(
+			ctx.rand, sampleAll(ctx.hparams, ctx.rand), model.TrialWorkloadSequencerType)
+		s.TrialRungs[create.RequestID] = 0
+		ops = append(ops, create)
+		ops = append(ops, NewTrain(create.RequestID, s.Rungs[0].UnitsNeeded))
+		ops = append(ops, NewValidate(create.RequestID))
+	}
+
+	return ops
+}
+
+func (s *asyncHalvingStoppingSearch) progress(float64) float64 {
+	allTrials := len(s.Rungs[0].Metrics)
+	// Give ourselves an overhead of 20% of maxTrials when calculating progress.
+	progress := float64(allTrials) / (1.2 * float64(s.MaxTrials))
+	if allTrials == s.MaxTrials {
+		numValidTrials := float64(s.TrialsCompleted) - float64(s.InvalidTrials)
+		progressNoOverhead := numValidTrials / float64(s.MaxTrials)
+		progress = math.Max(progressNoOverhead, progress)
+	}
+	return progress
+}
+
+func (s *asyncHalvingStoppingSearch) trialExitedEarly(
+	ctx context, requestID model.RequestID, exitedReason workload.ExitedReason,
+) ([]Operation, error) {
+	if exitedReason == workload.InvalidHP {
+		var ops []Operation
+		s.EarlyExitTrials[requestID] = true
+		ops = append(ops, NewClose(requestID))
+		s.ClosedTrials[requestID] = true
+		s.InvalidTrials++
+		// Remove metrics associated with InvalidHP trial across all rungs
+		highestRungIndex := s.TrialRungs[requestID]
+		for rungIndex := 0; rungIndex <= highestRungIndex; rungIndex++ {
+			rung := s.Rungs[rungIndex]
+			for i, trialMetric := range rung.Metrics {
+				if trialMetric.RequestID == requestID {
+					rung.Metrics = append(rung.Metrics[:i], rung.Metrics[i+1:]...)
+				}
+				break
+			}
```

Comment:
wouldn't this only remove the metric from the rung if it was the first one, otherwise it would break? it seems like the break is in the wrong place.

Reply:
good catch.  looks like this bug is in asha.go as well.  do you think i should fix asha.go in this PR or have a separate PR for that?